### PR TITLE
macOS Mouse Interface

### LINKED ISSUE #100: project doesn't build on macos
i have a macbookpro M1 from 2020, macos version is "monterey 12.3.1 (21E258)". i tried the two types of build, both failed.
here's what i got:

build type 1 (single binary)
```
mkdir build
cd build
cmake ..
make

results in:
[ build ]                                                                             ⍉
> pwd
/Users/klaus/Sources/dolphin/build

[ build ]
> make
[  0%] Built target cpp-optparse
[  1%] Built target imgui
[  4%] Built target glslang
[  4%] Performing update step for 'MoltenVK'
[  4%] Performing patch step for 'MoltenVK'
HEAD is now at b3ef160c Merge pull request #1470 from billhollings/sdk-1.2.198
[  4%] Performing configure step for 'MoltenVK'
CMake Error at /Users/klaus/Sources/dolphin/build/Externals/MoltenVK/MoltenVK-prefix/src/MoltenVK-stamp/MoltenVK-configure-Release.cmake:37 (message):
  Command failed: 127

   '/Users/klaus/Sources/dolphin/Externals/MoltenVK/configure.sh' '/Users/klaus/Sources/dolphin/build/Externals/MoltenVK/MoltenVK-prefix/src/MoltenVK-stamp' '/Users/klaus/Sources/dolphin/build/Externals/MoltenVK/MoltenVK-prefix/src/MoltenVK' 'v1.1.6'

  See also

    /Users/klaus/Sources/dolphin/build/Externals/MoltenVK/MoltenVK-prefix/src/MoltenVK-stamp/MoltenVK-configure-*.log


-- stdout output is:

========== Retrieving MoltenVK dependencies into External ==========

========== Cereal ==========

cereal repo: https://github.com/USCiLab/cereal.git
cereal revision: 51cbda5f30e56c801c07fe3d3aba5d7fb9e6cca4
Fetching origin

========== Vulkan-Headers ==========

Vulkan-Headers repo: https://github.com/KhronosGroup/Vulkan-Headers.git
Vulkan-Headers revision: 8c1c27d5a9b9de8a17f500053bd08c7ca6bba19c
Fetching origin

========== SPIRV-Cross ==========

SPIRV-Cross repo: https://github.com/KhronosGroup/SPIRV-Cross.git
SPIRV-Cross revision: 7c3cb0b12c9965497b08403c82ac1b82846fa7be
Fetching origin

========== glslang and SPIRV-Tools ==========

glslang repo: https://github.com/KhronosGroup/glslang.git
glslang revision: c9706bdda0ac22b9856f1aa8261e5b9e15cd20c5
Fetching origin

-- stderr output is:
cat: /Users/klaus/Sources/dolphin/build/Externals/MoltenVK/MoltenVK-prefix/src/MoltenVK-stamp/MoltenVK-last-version.txt: No such file or directory
HEAD is now at 51cbda5f Merge branch 'develop' for release 1.2.2
HEAD is now at 8c1c27d Update for Vulkan-Docs 1.2.198
HEAD is now at 7c3cb0b1 Merge pull request #1810 from billhollings/early-frag-tests-depth-out
HEAD is now at c9706bdd Merge pull request #2817 from greg-lunarg/sdk0
env: python: No such file or directory

CMake Error at /Users/klaus/Sources/dolphin/build/Externals/MoltenVK/MoltenVK-prefix/src/MoltenVK-stamp/MoltenVK-configure-Release.cmake:47 (message):
  Stopping after outputting logs.


make[2]: *** [Externals/MoltenVK/MoltenVK-prefix/src/MoltenVK-stamp/MoltenVK-configure] Error 1
make[1]: *** [Externals/MoltenVK/CMakeFiles/MoltenVK.dir/all] Error 2
make: *** [all] Error 2

[ build ]                                                                             ⍉
>
```

build type 2 (universal)
```
[ 61%] Building CXX object Source/Core/Core/CMakeFiles/core.dir/PowerPC/Jit64/Jit_Branch.cpp.o
[ 61%] Building CXX object Source/Core/Core/CMakeFiles/core.dir/PowerPC/Jit64/Jit_LoadStorePaired.cpp.o
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:124:11: warning: 10 enumeration values not handled in switch: 'INVALID_GAME', 'MENU', 'MENU_PRIME_1'... [-Wswitch]
  switch (game) {
          ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:124:11: note: add missing switch cases
  switch (game) {
          ^
[ 61%] Building CXX object Source/Core/Core/CMakeFiles/core.dir/PowerPC/Jit64/Jit_Paired.cpp.o
[ 62%] Building CXX object Source/Core/Core/CMakeFiles/core.dir/PowerPC/Jit64/Jit_SystemRegisters.cpp.o
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:380:20: error: 'path' is unavailable: introduced in macOS 10.15
  std::filesystem::path search_name(str);
                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:909:24: note: 'path' has been explicitly marked unavailable here
class _LIBCPP_TYPE_VIS path {
                       ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:380:25: error: 'path' is unavailable: introduced in macOS 10.15
  std::filesystem::path search_name(str);
                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:948:3: note: 'path' has been explicitly marked unavailable here
  path(const _Source& __src, format = format::auto_format) {
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:380:25: error: '~path' is unavailable: introduced in macOS 10.15
  std::filesystem::path search_name(str);
                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:968:3: note: '~path' has been explicitly marked unavailable here
  ~path() = default;
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:382:19: error: 'is_absolute' is unavailable: introduced in macOS 10.15
  if (search_name.is_absolute() && std::filesystem::exists(str)) {
                  ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:1406:34: note: 'is_absolute' has been explicitly marked unavailable here
  _LIBCPP_INLINE_VISIBILITY bool is_absolute() const {
                                 ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:382:53: error: 'exists' is unavailable: introduced in macOS 10.15
  if (search_name.is_absolute() && std::filesystem::exists(str)) {
                                                    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:1986:39: note: 'exists' has been explicitly marked unavailable here
inline _LIBCPP_INLINE_VISIBILITY bool exists(const path& __p) {
                                      ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:382:60: error: 'path' is unavailable: introduced in macOS 10.15
  if (search_name.is_absolute() && std::filesystem::exists(str)) {
                                                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:948:3: note: 'path' has been explicitly marked unavailable here
  path(const _Source& __src, format = format::auto_format) {
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:382:60: error: '~path' is unavailable: introduced in macOS 10.15
  if (search_name.is_absolute() && std::filesystem::exists(str)) {
                                                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:968:3: note: '~path' has been explicitly marked unavailable here
  ~path() = default;
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:35: error: 'absolute' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
                                  ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:1846:39: note: 'absolute' has been explicitly marked unavailable here
inline _LIBCPP_INLINE_VISIBILITY path absolute(const path& __p) {
                                      ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:44: error: 'path' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
                                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:948:3: note: 'path' has been explicitly marked unavailable here
  path(const _Source& __src, format = format::auto_format) {
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:44: error: '~path' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
                                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:968:3: note: '~path' has been explicitly marked unavailable here
  ~path() = default;
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:18: error: '~path' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
                 ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:968:3: note: '~path' has been explicitly marked unavailable here
  ~path() = default;
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:54: error: 'parent_path' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
                                                     ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:1366:34: note: 'parent_path' has been explicitly marked unavailable here
  _LIBCPP_INLINE_VISIBILITY path parent_path() const {
                                 ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:54: error: '~path' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
                                                     ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:968:3: note: '~path' has been explicitly marked unavailable here
  ~path() = default;
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:71: error: 'path' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
                                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:948:3: note: 'path' has been explicitly marked unavailable here
  path(const _Source& __src, format = format::auto_format) {
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:71: error: '~path' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
                                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:968:3: note: '~path' has been explicitly marked unavailable here
  ~path() = default;
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:69: error: 'operator/' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
                                                                    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:1500:41: note: 'operator/' has been explicitly marked unavailable here
  friend _LIBCPP_INLINE_VISIBILITY path operator/(const path& __lhs,
                                        ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:69: error: '~path' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
                                                                    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:968:3: note: '~path' has been explicitly marked unavailable here
  ~path() = default;
  ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:385:15: error: 'operator=' is unavailable: introduced in macOS 10.15
  search_name = (std::filesystem::absolute(rel_file).parent_path()) / str;
              ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:978:9: note: 'operator=' has been explicitly marked unavailable here
  path& operator=(path&& __p) noexcept {
        ^
/Users/klaus/Sources/dolphin/Source/Core/Core/PrimeHack/Mods/ElfModLoader.cpp:387:24: error: 'exists' is unavailable: introduced in macOS 10.15
  if (std::filesystem::exists(search_name)) {
                       ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/filesystem:1986:39: note: 'exists' has been explicitly marked unavailable here
inline _LIBCPP_INLINE_VISIBILITY bool exists(const path& __p) {
                                      ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
1 warning and 20 errors generated.
make[2]: *** [Source/Core/Core/CMakeFiles/core.dir/PrimeHack/Mods/ElfModLoader.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Source/Core/Core/CMakeFiles/core.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
-- MoltenVK build command succeeded.  See also /Users/klaus/Sources/dolphin/build/x86_64/Externals/MoltenVK/MoltenVK-prefix/src/MoltenVK-stamp/MoltenVK-build-*.log
[ 62%] No install step for 'MoltenVK'
[ 62%] Completed 'MoltenVK'
[ 62%] Built target MoltenVK
make: *** [all] Error 2
Traceback (most recent call last):
  File "/Users/klaus/Sources/dolphin/build/../BuildMacOSUniversalBinary.py", line 367, in <module>
    build(conf)
  File "/Users/klaus/Sources/dolphin/build/../BuildMacOSUniversalBinary.py", line 298, in build
    subprocess.check_call(["cmake", "--build", ".",
  File "/opt/homebrew/Cellar/python@3.9/3.9.12/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--config', 'Release', '--parallel', '8']' returned non-zero exit status 2.

[ build ]                                                                             ⍉
>
```
i've installed qt5 from brew (`brew install qt@5`)
i do not have python2, and available python3 seems to match with `../BuildMacOSUniversalBinary.py` which have header:
```
!/usr/bin/env python3
```
both build seems to fail at moltenvk build step.

### PR DESCRIPTION
**Pull request by brandonsorensen**
Adds a `GenericMouse` implementation called `QuartzInputMouse` as well as some minor modifications to Dolphin's build instructions that enable building PrimeHack on macOS. I don't have a ton of experience in C++, so hopefully it's a stellar enough implementation. It lifts a lot of code from existing Dolphin interfaces.

Credit to the conversation in #100, which got me on the right track and which this PR should resolve #100.
 

So far I have only tested on two machines:

- M1 Mac mini
- Intel MacBook Pro 2019

Both machines were running macOS 12.4. The minimum build target is macOS 11.0. As such, it should not build on anything running an OS older than macOS Big Sur.

Testing was minimal. I only played though the first bits of _Echoes_ and _Corruption_ on a _Trilogy_ ROM.

I was unable to build a universal binary (i.e., a binary that runs on both M1 and Intel machines). Frankly that process sounds like quite a headache. Hopefully someone could take up that task in the future.

It should be noted that the optimal performance settings described in PrimeHack's Wiki do not apply to macOS, or at least the M1 hardware I tested on. In particular, turning off **Synchronize GPU thread** led to constant crashes. YMMV, of course.

The only change that I believe has any chance of affecting non-macOS builds is at `Source/Core/DolphinQt/CMakeLists.txt`. Following advice from [this thread](https://bugs.dolphin-emu.org/issues/11332), I removed some lines from build instructions on Apple devices that seemed to link multiple versions of Qt. It shouldn't affect any other platforms, but I have not tested it on non-Apple devices to confirm.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -12,7 +12,7 @@ cmake_minimum_required(VERSION 3.10)
 # This is done in the universal binary building script to build a binary that 
 # runs on 10.13 on x86_64 computers, while still containing an arm64 slice.
 
-set(CMAKE_OSX_DEPLOYMENT_TARGET "10.14.0" CACHE STRING "")
+ set(CMAKE_OSX_DEPLOYMENT_TARGET "11.0" CACHE STRING "")
```
Comment (vyuuui) [inline CMakeLists.txt:15]: Nit: added whitespace

File: Source/Core/DolphinQt/CMakeLists.txt [hunk 2]
```diff
@@ -558,9 +558,11 @@ if(APPLE)
...
   endif()
 
-  # Update library references to make the bundle portable
-  include(DolphinPostprocessBundle)
-  dolphin_postprocess_bundle(dolphin-emu)
+  if(NOT APPLE)
+    # Update library references to make the bundle portable
+    include(DolphinPostprocessBundle)
+    dolphin_postprocess_bundle(dolphin-emu)
+ endif()
```
Comment (vyuuui) [inline Source/Core/DolphinQt/CMakeLists.txt:565]: Nit: wrong tabbing

File: Source/Core/InputCommon/ControllerInterface/Quartz/QuartzKeyboardAndMouse.mm [hunk 3]
```diff
@@ -141,7 +142,7 @@
   // keys that aren't being recognized, bump this number up!
   for (int keycode = 0; keycode < 0x80; ++keycode)
     AddInput(new Key(keycode));
-
+
```
Comment (vyuuui) [inline Source/Core/InputCommon/ControllerInterface/Quartz/QuartzKeyboardAndMouse.mm:145]: Nit: unnecessary change

File: Source/Core/InputCommon/ControllerInterface/Quartz/QuartzKeyboardAndMouse.mm [hunk 4]
```diff
@@ -175,6 +177,7 @@
 {
   CGRect bounds = CGRectZero;
   CGWindowID windowid[1] = {m_windowid};
+
```
Comment (vyuuui) [inline Source/Core/InputCommon/ControllerInterface/Quartz/QuartzKeyboardAndMouse.mm:180]: Nit: unnecessary change

File: Source/Core/InputCommon/QuartzInputMouse.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
+#include <ApplicationServices/ApplicationServices.h>
+#include "GenericMouse.h"
+
+namespace prime
+{
+
+bool InitCocoaInputMouse(uint32_t* windowid);
+CGRect getBounds(uint32_t* m_windowid);
+CGPoint getWindowCenter(uint32_t* m_windowid);
```
Comment (vyuuui) [inline Source/Core/InputCommon/QuartzInputMouse.h:9]: `getBounds` and `getWindowCenter` don't need to be exposed outside of `QuartzInputMouse.mm`, so they shouldn't be declared here. Mark them as static.

**Comment by vyuuui** [comment]
Looks good, I'll check if this affects the linux/windows builds at all.

**Comment by vyuuui** [comment]
Seems like a recent update to libstdc++ has broken a transitive include on linux, but otherwise it builds fine on linux & windows. It's nothing for you to worry about.

**Comment by vyuuui** [comment]
Sorry for the late response.
Just to reduce the commit volume, can you squash those fixes together? After that, it's good to merge.

**Comment by brandonsorensen** [comment]
I squashed the last five containing small formatting fixes together and left the one that moved those helper functions into the mouse class's scope. I can squash that one as well, but it seems substantial enough to warrant its own commit, if it's all the same to you. Otherwise, should be good to go!

**Comment by vyuuui** [comment]
Looks good, this should be tested by the reporter of #100  to validate.
